src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -0,0 +1,544 @@
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using SixLabors.ImageSharp.Memory;
+
+namespace SixLabors.ImageSharp.IO
+{
+    /// <summary>
+    /// Provides an in-memory stream composed of non-contiguous chunks that doesn't need to be resized.
+    /// Chunks are allocated by the <see cref="MemoryAllocator"/> assigned via the constructor
+    /// and is designed to take advantage of buffer pooling when available.
+    /// </summary>
+    internal sealed class ChunkedMemoryStream : Stream
+    {
+        /// <summary>
+        /// The default length in bytes of each buffer chunk.
+        /// </summary>
+        public const int DefaultBufferLength = 4096;
+
+        // The memory allocator.
+        private readonly MemoryAllocator allocator;
+
+        // Data
+        private MemoryChunk memoryChunk;
+
+        // The length of each buffer chunk
+        private readonly int chunkLength;
+
+        // Has the stream been disposed.
+        private bool isDisposed;
+
+        // Current chunk to write to
+        private MemoryChunk writeChunk;
+
+        // Offset into chunk to write to
+        private int writeOffset;
+
+        // Current chunk to read from
+        private MemoryChunk readChunk;
+
+        // Offset into chunk to read from
+        private int readOffset;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ChunkedMemoryStream"/> class.
+        /// </summary>
+        public ChunkedMemoryStream(MemoryAllocator allocator)
+            : this(DefaultBufferLength, allocator)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ChunkedMemoryStream"/> class.
+        /// </summary>
+        /// <param name="bufferLength">The length, in bytes of each buffer chunk.</param>
+        /// <param name="allocator">The memory allocator.</param>
+        public ChunkedMemoryStream(int bufferLength, MemoryAllocator allocator)
+        {
+            Guard.MustBeGreaterThan(bufferLength, 0, nameof(bufferLength));
+            Guard.NotNull(allocator, nameof(allocator));
+
+            this.chunkLength = bufferLength;
+            this.allocator = allocator;
+        }
+
+        /// <inheritdoc/>
+        public override bool CanRead => !this.isDisposed;
+
+        /// <inheritdoc/>
+        public override bool CanSeek => !this.isDisposed;
+
+        /// <inheritdoc/>
+        public override bool CanWrite => !this.isDisposed;
+
+        /// <inheritdoc/>
+        public override long Length
+        {
+            get
+            {
+                this.EnsureNotDisposed();
+
+                int length = 0;
+                MemoryChunk chunk = this.memoryChunk;
+                while (chunk != null)
+                {
+                    MemoryChunk next = chunk.Next;
+                    if (next != null)
+                    {
+                        length += chunk.Length;
+                    }
```

Comment:
If we store the number of "fully written" chunks, the loop could be replaced by a multiplication.

Reply:
I had a look but the benchmarks showed no need for the optimization so chose not to complicate things.